onnxruntime/core/framework/plugin_data_transfer.cc

```diff
@@ -0,0 +1,68 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+#include "core/framework/plugin_data_transfer.h"
+
+#include "core/framework/error_code_helper.h"
+
+namespace onnxruntime {
+namespace plugin_ep {
+
+namespace {
+static const std::function<void(void*)> no_op_deleter = [](void*) {};
+static const MLDataType ml_tensor_type = DataTypeImpl::GetType<Tensor>();
+}  // namespace
+
+Status DataTransfer::CopyTensors(const std::vector<SrcDstPair>& src_dst_pairs) const {
+  // need to wrap the src/dst Tensor instances in OrtValue as the ORT API doesn't expose an OrtTensor.
+  // Adding an OrtTensor to the API would also require adding getters for type/shape/data.
+  // Those already exist for OrtValue so in order to minimize the API surface area we pay the price of a
+  // const_cast to convert the `const Tensor*` src to an OrtValue.
+  std::vector<OrtValue> values;
+  values.resize(src_dst_pairs.size() * 2);
+
+  for (size_t i = 0; i < src_dst_pairs.size(); ++i) {
+    const auto& pair = src_dst_pairs[i];
+
+    // we need to remove the const from the src to wrap it in an OrtValue.
+    // it's passed to the impl as a const OrtValue, and the deleter is a no-op so this should be safe.
+    Tensor* src_tensor = const_cast<Tensor*>(&(pair.src.get()));
+    values[i * 2].Init(reinterpret_cast<void*>(src_tensor), ml_tensor_type, no_op_deleter);
```

Comment:
please feel free to ignore this as this const_cast is safe.

If storing unowned data into OrtValue becomes common, we could create a new overload of `OrtValue::Init()` that does this internally and knows to use a no-op deleter.

```c++
  void Init(const void* pUnownedData, onnxruntime::MLDataType type) {
    const std::function<void(void*)> no_op_deleter = [](void*) {};  // could be a static variable
    data_.reset(const_cast<void*>(pUnownedData), no_op_deleter);
    type_ = type;
  }
```